controller/api/destination/watcher/pod_watcher.go

```diff
@@ -0,0 +1,411 @@
+		AddFunc:    pw.addPod,
+		DeleteFunc: pw.deletePod,
+		UpdateFunc: pw.updatePod,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = k8sAPI.Srv().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    pw.updateServer,
+		DeleteFunc: pw.updateServer,
+		UpdateFunc: func(_, obj interface{}) { pw.updateServer(obj) },
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return pw, nil
+}
+
+// Subscribe notifies the listener on changes on any pod backing the passed
+// ip:port or its associated opaque protocol config
+func (pw *PodWatcher) Subscribe(pod *corev1.Pod, ip string, port Port, listener PodUpdateListener) {
+	pw.log.Debugf("Establishing watch on %s:%d", ip, port)
+	pp := pw.getOrNewPodPublisher(pod, ip, port)
+	pp.subscribe(listener)
+}
+
+// Subscribe stops notifying the listener on chages on any pod backing the
+// passed ip:port or its associated protocol config
+func (pw *PodWatcher) Unsubscribe(ip string, port Port, listener PodUpdateListener) {
+	pw.mu.Lock()
+	defer pw.mu.Unlock()
+
+	pw.log.Debugf("Stopping watch on %s:%d", ip, port)
+	pp, ok := pw.getPodPublisher(ip, port)
+	if !ok {
+		pw.log.Errorf("Cannot unsubscribe from unknown ip:port [%s:%d]", ip, port)
+		return
+	}
+	pp.unsubscribe(listener)
+
+	if len(pp.listeners) == 0 {
+		delete(pw.publishers, IPPort{pp.ip, pp.port})
+	}
+}
+
+// addPod is an event handler so it cannot block
+func (pw *PodWatcher) addPod(obj any) {
+	pod := obj.(*corev1.Pod)
+	pw.log.Tracef("Added pod %s.%s", pod.Name, pod.Namespace)
+	go pw.submitPodUpdate(pod, false)
+}
+
+// deletePod is an event handler so it cannot block
+func (pw *PodWatcher) deletePod(obj any) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			pw.log.Errorf("Couldn't get object from DeletedFinalStateUnknown %#v", obj)
+			return
+		}
+		pod, ok = tombstone.Obj.(*corev1.Pod)
+		if !ok {
+			pw.log.Errorf("DeletedFinalStateUnknown contained object that is not a Pod %#v", obj)
+			return
+		}
+	}
+	pw.log.Tracef("Deleted pod %s.%s", pod.Name, pod.Namespace)
+	go pw.submitPodUpdate(pod, true)
+}
+
+// updatePod is an event handler so it cannot block
+func (pw *PodWatcher) updatePod(oldObj any, newObj any) {
+	oldPod := oldObj.(*corev1.Pod)
+	newPod := newObj.(*corev1.Pod)
+	if oldPod.DeletionTimestamp == nil && newPod.DeletionTimestamp != nil {
+		// this is just a mark, wait for actual deletion event
+		return
+	}
+	pw.log.Tracef("Updated pod %s.%s", newPod.Name, newPod.Namespace)
+	go pw.submitPodUpdate(newPod, false)
+}
+
+func (pw *PodWatcher) submitPodUpdate(pod *corev1.Pod, remove bool) {
+	pw.mu.RLock()
+	defer pw.mu.RUnlock()
+
+	submitPod := pod
+	if remove {
+		submitPod = nil
+	}
+	for _, container := range pod.Spec.Containers {
+		for _, containerPort := range container.Ports {
+			if containerPort.ContainerPort != 0 {
+				if pp, ok := pw.getPodPublisher(pod.Status.PodIP, Port(containerPort.ContainerPort)); ok {
```

Comment:
Should we be using `getOrNewPodPublisher` here instead so that we're creating a podPublisher with no listeners?  We are, effectively, building our own index.

Reply:
Not sure I follow. Of what use are podPublishers with no listeners? Sounds like a waste of resources to keep podPublishers for all pods in the cluster. If some client becomes interested in the given ip:port, then the `Subscribe()` method will make sure the create the podPublisher (using the pod passed to `Subscribe()` or retrieving it itself as per your previous comment).

Reply:
Looking at other watchers like the endpointsWatcher or the profileWatcher, it looks like we populate the publisher map when either a listener subscribes OR when we get an update for that object.  But you're right, we can build the full state when a subscription happens.